### PR TITLE
✨ Allow to perform manual ETCD snapshots

### LINKED ISSUE #725: Implement snapshot controller functionality
We now have the skeleton for the controller, we need to implement actual making of snapshots using system-agent

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
**What this PR does / why we need it**:

This PR completes the initial setup of the controller with functionality to create manual snapshots on user demand, using the `system-agent` and RKE2 cli, following the [docs](https://docs.rke2.io/backup_restore?_highlight=snapsho#creating-snapshots).

This comes with some problems, like inability to specifically name the snapshot and deterministically connect it to the `ETCDMachineSnapshot` replica, created by the user, but the issue lays in lacking options in `k3s/rke2` CLI. In the scenario when the snapshot is named “my-snapshot”, the actual name looks like: `local-my-snapshot-rke2-control-plane-mnf8z-1731675906-31aa01` and may easily overlap with a snapshot named as just “snapshot”. (Edit: currently matches against `snapshotName` and `machineName`)

This also makes it hard to connect the `creationTimestamp` of the snapshot on the child cluster with the originally invoked `ETCDMachineSnapshot` resource execution, as no additional info can be stored on the snapshot instance when calling the CLI. (Edit: storing snapshot timestamp in the list now)

Final snapshot location and other infor is created and is available on the list within the automatic snapshot record (one per cluster).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #725

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits into logical changes
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

File: exp/etcdrestore/controllers/etcdmachinesnapshot_controller.go [hunk 1]
```diff
@@ -110,52 +125,101 @@ func (r *ETCDMachineSnapshotReconciler) Reconcile(ctx context.Context, req ctrl.
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to collect machines for cluster: %w", err)
+	}
+
+	controlPlaneMachines := machines.Filter(collections.ControlPlaneMachines(cluster.Name))
+	targetMachineCandidates := controlPlaneMachines.Filter(func(machine *clusterv1.Machine) bool {
+		return machine.Name == etcdMachineSnapshot.Spec.MachineName
+	}).UnsortedList()
+
+	if len(targetMachineCandidates) < 1 {
```
Comment (alexander-demicev) [inline exp/etcdrestore/controllers/etcdmachinesnapshot_controller.go:151]: should it be `!=` ?

**Review by alexander-demicev — DISMISSED** [review]
thanks! looks good to me

File: exp/etcdrestore/controllers/etcdmachinesnapshot_controller.go (hunk 1 above)
Comment (Danil-Grigorev) [reply exp/etcdrestore/controllers/etcdmachinesnapshot_controller.go:151]: It makes little difference, current assumption is that the first in the list should be used.

**Review by alexander-demicev — APPROVED** [review]

**Review by yiannistri — APPROVED** [review]
LGTM

**Review by furkatgofurov7 — APPROVED** [review]

**Pull request merged**